ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Examples.hs

```diff
@@ -0,0 +1,377 @@
+  where
+    serverIdle :: forall (n :: N).
+                  [tx]
+               -> Nat n
+               -> ServerState txid tx
+               -> ServerStIdle n txid tx m [tx]
+    serverIdle accum Zero st
+        -- There are no replies in flight, but we do know some more txs we can
+        -- ask for, so lets ask for them and more txids.
+      | canRequestMoreTxs st
+      = serverReqTxs accum Zero st
+
+        -- There's no replies in flight, and we have no more txs we can ask for
+        -- so the only remaining thing to do is to ask for more txids. Since
+        -- this is the only thing to do now, we make this a blocking call.
+      | otherwise
+      , let numTxIdsToRequest = maxTxIdsToRequest `min` maxUnacked
+      = assert (requestedTxIdsInFlight st == 0
+             && Seq.null (unacknowledgedTxIds st)
+             && Map.null (availableTxids st)
+             && Map.null (bufferedTxs st)) $
+        SendMsgRequestTxIdsBlocking
+          (numTxsToAcknowledge st)
+          numTxIdsToRequest
+          accum                      -- result if the client reports we're done
+          (\txids -> do
+              traceWith tracer (EventRequestTxIdsBlocking st (numTxsToAcknowledge st) numTxIdsToRequest)
+              return . handleReply accum Zero st {
+                         numTxsToAcknowledge    = 0,
+                         requestedTxIdsInFlight = numTxIdsToRequest
+                       }
+                     . CollectTxIds numTxIdsToRequest
+                     . NonEmpty.toList $ txids)
+
+    serverIdle accum (Succ n) st
+        -- We have replies in flight and we should eagerly collect them if
+        -- available, but there are transactions to request too so we should
+        -- not block waiting for replies.
+        --
+        -- Having requested more transactions, we opportunistically ask for
+        -- more txids in a non-blocking way. This is how we pipeline asking for
+        -- both txs and txids.
+        --
+        -- It's important not to pipeline more requests for txids when we have
+        -- no txs to ask for, since (with no other guard) this will put us into
+        -- a busy-polling loop.
+        --
+      | canRequestMoreTxs st
+      = CollectPipelined
+          (Just (serverReqTxs accum (Succ n) st))
+          (handleReply accum n st)
+
+        -- In this case there is nothing else to do so we block until we
+        -- collect a reply.
+      | otherwise
+      = CollectPipelined
+          Nothing
+          (handleReply accum n st)
+
+    canRequestMoreTxs :: ServerState k tx -> Bool
+    canRequestMoreTxs st =
+        not (Map.null (availableTxids st))
+
+    handleReply :: forall (n :: N).
+                   [tx]
+                -> Nat n
+                -> ServerState txid tx
+                -> Collect txid tx
+                -> ServerStIdle n txid tx m [tx]
+    handleReply accum n st (CollectTxIds reqNo txids) =
```

Comment:
We are missing a monadic action here which will allow to place the received txs in the mempool.

Reply:
Yes, that was what the other TODO was about, and is resolved in a subsequent PR for the proper versions (for exactly that reason).